model/project_vars.go

```diff
@@ -296,6 +361,228 @@ func (projectVars *ProjectVars) Upsert() (*adb.ChangeInfo, error) {
+
+	for varName, varValue := range varsToUpsert {
+		partialParamName, paramValue, err := convertVarToParam(projectID, pm, varName, varValue)
+		if err != nil {
+			catcher.Wrapf(err, "converting project variable '%s' to parameter", varName)
+			continue
+		}
+		param, err := paramMgr.Put(ctx, partialParamName, paramValue)
+		if err != nil {
+			catcher.Wrapf(err, "putting project variable '%s' into Parameter Store", varName)
+			continue
+		}
+		paramName := param.Name
+
+		paramMappingsToUpsert[varName] = ParameterMapping{
+			Name:          varName,
+			ParameterName: paramName,
+		}
+
+		if existingParamMapping, ok := nameToExistingParamMapping[varName]; ok && existingParamMapping.ParameterName != paramName {
+			// In a few special edge cases, the project var could already be
+			// stored in one parameter name but has to be renamed to a new
+			// parameter. For example, if the project var is stored in a
+			// parameter named "foo" initially and then the value is updated a
+			// very long string, the parameter could be renamed to "foo.gz" to
+			// indicate that it had to be compressed to fit within the parameter
+			// 8 KB limitation. If the parameter has been renamed, then the old
+			// parameter name is now invalid and should be cleaned up.
+			if err := paramMgr.Delete(ctx, existingParamMapping.ParameterName); err != nil {
+				catcher.Wrapf(err, "deleting project variable '%s' from Parameter Store whose parameter was renamed from '%s' to '%s'", varName, existingParamMapping.ParameterName, paramName)
+				continue
+			}
+		}
+	}
+	return paramMappingsToUpsert, nil
+}
+
+// deleteParameters deletes parameters corresponding to deleted project variables
+// from Parameter Store. It returns the parameter mappings for the deleted
+// variables.
+func (projectVars *ProjectVars) deleteParameters(ctx context.Context, pm ParameterMappings, varsToDelete map[string]struct{}) (map[string]ParameterMapping, error) {
+	nameToExistingParamMapping := pm.NameMap()
+	paramMappingsToDelete := make(map[string]ParameterMapping, len(varsToDelete))
+	for varToDelete := range varsToDelete {
+		if paramMapping, ok := nameToExistingParamMapping[varToDelete]; ok {
+			paramMappingsToDelete[varToDelete] = paramMapping
+		}
+	}
+
+	namesToDelete := make([]string, 0, len(paramMappingsToDelete))
+	for _, m := range paramMappingsToDelete {
+		namesToDelete = append(namesToDelete, m.ParameterName)
+	}
+
+	if len(namesToDelete) > 0 {
+		paramMgr := evergreen.GetEnvironment().ParameterManager()
+		if err := paramMgr.Delete(ctx, namesToDelete...); err != nil {
+			return nil, err
+		}
+	}
+
+	return paramMappingsToDelete, nil
+}
+
+// getProjectVarsDiff returns the diff of added/updated/deleted project
+// variables between the before and after project variables. It returns the
+// variables that have to be upserted and deleted so that before matches after.
+func getProjectVarsDiff(before, after *ProjectVars) (upserted map[string]string, deleted map[string]struct{}) {
```

Comment:
We don't use the lookup for the deleted map I think, we only loop through the keys. It's not a noticable performance difference but should we just convert it to a slice? Or do you foresee using lookup?